### PR TITLE
adding support for configmap in mutation reconciler

### PR DESCRIPTION
**Pull request by souleb**
## Description

If implemented, adds a `configMapSource` that can be used to retrieve a `values.yaml` in order to perform config substituons.

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [x] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

- Related Issue # (issue)
- Closes # (issue)
- Fixes # (issue)
> Remove if not applicable

## Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help
- [ ] Separate ticket for tests # (issue/pr)

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Added to documentation?

- [ ] 📜 README.md
- [ ] 🙅 no documentation needed

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: api/v1alpha1/mutation_types.go [hunk 1]
```diff
@@ -50,12 +50,22 @@ type MutationSpec struct {
 type ValuesSource struct {
 	// +optional
 	FluxSource *FluxValuesSource `json:"fluxSource,omitempty"`
-
-	// TODO
-	// ConfigMap meta.LocalObjectReference
+	// +optional
+	ConfigMapSource *ConfigMapSource `json:"configMapSource,omitempty"`
+	// +optional
+	Key string `json:"key,omitempty"`
```
Comment (Skarlso) [inline api/v1alpha1/mutation_types.go:56]: Wasn't this key supposed to be the key inside the ConfigMapSource? Maybe it's a remnant of an older implementation?

File: api/v1alpha1/mutation_types.go [hunk 2]
```diff
@@ -50,12 +50,22 @@ type MutationSpec struct {
...
 	// Secret meta.LocalObjectReference
 }
 
+type ConfigMapSource struct {
+	// +required
+	SourceRef meta.LocalObjectReference `json:"sourceRef,omitempty"`
+	// +required
+	Key string `json:"key,omitempty"`
```
Comment (Skarlso) [inline api/v1alpha1/mutation_types.go:64]: If required, remove `omitempty`. Same for subPath.

File: controllers/mutation_reconcile_looper.go [hunk 3]
```diff
@@ -15,6 +15,8 @@ import (
 	"path/filepath"
 	"strings"
 
+	corev1 "k8s.io/api/core/v1"
```
Comment (Skarlso) [inline controllers/mutation_reconcile_looper.go:18]: Can you put this with the rest of the imports, please?

File: controllers/mutation_reconcile_looper.go [hunk 4]
```diff
@@ -801,19 +804,47 @@ func (m *MutationReconcileLooper) getValues(ctx context.Context, obj *v1alpha1.M
...
 		}
-
 		var found bool
 		if obj.ValuesFrom.FluxSource.SubPath != "" {
 			data, found = extractSubpath(data, obj.ValuesFrom.FluxSource.SubPath)
 			if !found {
 				return nil, errors.New("subPath not found")
 			}
 		}
+	} else if obj.ValuesFrom.ConfigMapSource != nil {
```
Comment (Skarlso) [inline controllers/mutation_reconcile_looper.go:817]: Now I wonder if these could be defined together? Some values come from Config, others from FluxSource?

Comment (Skarlso) [reply controllers/mutation_reconcile_looper.go:817]: On a side-note. Can you please extract these into functions? This if/else block is now rather huge and hard to follow.

File: controllers/mutation_reconcile_looper.go [hunk 5]
```diff
@@ -801,19 +804,47 @@ func (m *MutationReconcileLooper) getValues(ctx context.Context, obj *v1alpha1.M
...
+		cm := &corev1.ConfigMap{}
+		key := types.NamespacedName{
+			Name:      obj.ValuesFrom.ConfigMapSource.SourceRef.Name,
+			Namespace: namespace,
```
Comment (Skarlso) [inline controllers/mutation_reconcile_looper.go:821]: In case of MPAS this would be the Project namespace. Is that where the ConfigMap would reside? Is it possible it's gonna be in a different namespace?

Comment (souleb) [reply controllers/mutation_reconcile_looper.go:821]: I think the values.yam is for a product for a given project. So it has to reside in the project ns.

Comment (Skarlso) [reply controllers/mutation_reconcile_looper.go]: Sounds good!

File: Makefile [hunk 6]
```diff
@@ -77,7 +77,7 @@ vet: ## Run go vet against code.
 
 .PHONY: test
 test: manifests generate fmt vet build-wasm-testdata envtest ## Run tests.
-	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path)" go test ./... -coverprofile cover.out
+	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path)" go test -v ./... -run TestConfigurationReconciler -coverprofile cover.out
```
Comment (Skarlso) [inline Makefile:80]: Shouldn't you remove this?

Comment (souleb) [reply Makefile:80]: oh thanks.

**Review by phoban01 — APPROVED** [review]
